### PR TITLE
Mark all graphnodes of a block as dirty on entry

### PR DESCRIPTION
**Pull request by junmendoza**
@ke-yu 

All nodes in a block of code is set to dirty on entry.
This is a general behavior applied to assoc update regardless of normal or live-execution.

This also fixes ScopeIf where the true and false body need to be executed multiple times for multiple modifications.
http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6322

### CODE REVIEW

File: src/Engine/ProtoCore/AssociativeGraph.cs [hunk 1]
```diff
@@ -69,11 +69,14 @@ public static int GetDirtyNodeCountAtGlobalScope(Executable exe)
...
+            if (graphNodes != null)
             {
-                if (graphNode.isDirty)
+                foreach (AssociativeGraph.GraphNode graphNode in graphNodes)
                 {
-                    ++dirtyNodes;
+                    if (graphNode.isDirty)
+                    {
+                        ++dirtyNodes;
+                    }
```
Comment (ke-yu) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: The only place where `GetDirtyNodeCountAtGlobalScope()` is used is in `ApplyUpdate()` in LiveRunner.cs:
```
if (ProtoCore.AssociativeEngine.Utils.GetDirtyNodeCountAtGlobalScope(runnerCore.DSExecutable) > 0)
{
    ...
}
```
So it is not necessary to go through all graph nodes to answer a "Yes/No" question.

File: src/Engine/ProtoCore/AssociativeGraph.cs [hunk 2]
```diff
@@ -1259,6 +1262,28 @@ public List<GraphNode> GraphList
...
+                foreach (GraphNode gnode in gnodeList)
+                {
+                    if (gnode.languageBlockId == block)
+                    {
+                        gnode.isDirty = true;
+                    }
+                }
+            }
+        }
+
```
Comment (ke-yu) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: This overloaded `MarkGraphNodesDirty()` implementation is quite different from other versions which checks
  - If a node is a binary expression
  - Check `OriginalAstID`
  - Set `GraphNode.IsActive`.

Are those extra checking and settings needed in this overloaded function as well?

File: src/Engine/ProtoCore/AssociativeGraph.cs (hunk 1 above)
Comment (junmendoza) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: Done, Ive replaced this with a simpler function that determines if at least one graphnode is dirty at the global scope.
The old function was implemetned for different uses but is now removed

File: src/Engine/ProtoCore/AssociativeGraph.cs (hunk 2 above)
Comment (junmendoza) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: Ive replaced the name to MarkAllGraphNodesDirty. It is meant to mark all the graphnodes in scope as dirty regardless of any condition

File: src/Engine/ProtoCore/AssociativeGraph.cs [hunk 3]
```diff
@@ -59,24 +59,27 @@ public static List<AssociativeGraph.GraphNode> GetGraphNodesFromAST(ProtoCore.DS
...
-        /// Gets the number of dirty VM graphnodes at the global scope
+        /// Determines if at least one graphnode in the glboal scope is dirty
         /// </summary>
         /// <param name="exe"></param>
         /// <returns></returns>
-        public static int GetDirtyNodeCountAtGlobalScope(Executable exe)
+        public static bool IsGlobalScopeDirty(Executable exe)
         {
             Validity.Assert(exe != null);
             int dirtyNodes = 0;
```
Comment (ke-yu) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: `dirtyNodes` could be removed now.

Comment (junmendoza) [inline src/Engine/ProtoCore/AssociativeGraph.cs]: Done

**Comment by ke-yu** [comment]
LGTM

**Pull request merged**